cql3: Allow for describing CDC log tables

dawmd opened the pull request:
In the past, DESC SCHEMA would produce create statements for both the base
and the log table. That was incorrect as the log table is automatically
created alongside the base one. That was solved in scylladb/scylladb@9ab57b1
(scylladb/scylladb#18467).

The mentioned changes implemented the following solution:

* DESC SCHEMA/KEYSPACE/TABLE would still print a create statement for the
  CDC base table,
* DESC SCHEMA/KEYSPACE would start printing an alter statement for the
  CDC log table. That statement would ensure that the restored log table
  has the same parameters as the original one,
* DESC TABLE <base table> would behave as DESC SCHEMA/KEYSPACE, i.e.
  it would print a create statement for the base table and an alter
  statement for the log table,
* DESC TABLE <log table> would result in an error.

While that solution was good and behaved correctly in the context of
restoring the schema, it had one flaw: describe statement aren't only
used as a means for producing a backup; they also serve an informative
purpose to learn about the schema, e.g. to learn what parameters a specific
table uses. Because we didn't allow for describing CDC log tables, the user
couldn't look them up directly via a describe statement -- they had to
describe the base table for that.

Attempting to describe a log table ended with an error, e.g.:

```
$ DESC TABLE ks.t_scylla_cdc_log;

ks.t_scylla_cdc_log is a cdc log table and it cannot be described directly. Try `DESC TABLE ks.t` to describe cdc base table and it's log table.
```

In these changes, we allow for describing CDC log tables again. The
semantics of the first three bullets above remains unchanged, but
we impose new behavior for DESC TABLE <log table>:

* When the user executes DESC TABLE <log table>, a create statement
  will be returned, treating the table as if it were a regular one,
* The create statement will be wrapped in CQL comment markers.

The rationale for the second bullet is that although we want to give the
user a means to look into the structure and options of a CDC log table,
the returned statement is not supposed to be ever executed by them. We
want to minimize the risk of that.

An example of the behavior after the change:

```
$ DESC TABLE ks.t_scylla_cdc_log;

/* Do NOT execute this statement! It's only for informational purposes.
   A CDC log table is created automatically when the base is created.

CREATE TABLE ks.t_scylla_cdc_log (
    "cdc$stream_id" blob,
    "cdc$time" timeuuid,
    "cdc$batch_seq_no" int,
    "cdc$end_of_batch" boolean,
    "cdc$operation" tinyint,
    "cdc$ttl" bigint,
    p int,
    PRIMARY KEY ("cdc$stream_id", "cdc$time", "cdc$batch_seq_no")
) WITH CLUSTERING ORDER BY ("cdc$time" ASC, "cdc$batch_seq_no" ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'enabled': 'false', 'keys': 'NONE', 'rows_per_partition': 'NONE'}
    AND comment = 'CDC log for ks.t'
    AND compaction = {'class': 'TimeWindowCompactionStrategy', 'compaction_window_size': '60', 'compaction_window_unit': 'MINUTES', 'expired_sstable_check_frequency_seconds': '1800'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 0
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND speculative_retry = '99.0PERCENTILE';

*/
```

We also extend the developer documentation regarding DESCRIBE
statements on CDC tables.

Fixes scylladb/scylladb#21235

Backport: these changes are an enhancement, so not needed.

nyh commented:
I'm confused about the rationale of this patch. You are claiming that you are fixing issue #13702, but as far as I understand (maybe I am misunderstanding...), that issue was about DESC TABLE on the **base table** - the user complained it doesn't work at all, and other commenters discussed what we expect its output to look like.

But it seems your patch is about allowing DESC TABLE the CDC **log table** by name. This is what I see described in the commit message and being tested in the tests. So it seems a different issue from #13702.

Finally, please include a short example in your cover letter. I was really confused on whether you're fixing DESC TABLE on the base table or the cdc log table, and what exactly the "before" and "after" outputs look like.

dawmd commented:
> I'm confused about the rationale of this patch. You are claiming that you are fixing issue #13702, but as far as I understand (maybe I am misunderstanding...), that issue was about DESC TABLE on the **base table** - the user complained it doesn't work at all, and other commenters discussed what we expect its output to look like.
>
> But it seems your patch is about allowing DESC TABLE the CDC **log table** by name. This is what I see described in the commit message and being tested in the tests. So it seems a different issue from #13702.

I agree the issue itself is quite confusing and chances are it would be better to close it and open a new one even if for the sake of clarity only. The reason why this patch references it is because it's part of the effort to resolve it, i.e. fix the interaction of describe statements with CDC tables. They already work as intended in most cases, the only thing missing being describing a CDC log table.

> Finally, please include a short example in your cover letter. I was really confused on whether you're fixing DESC TABLE on the base table or the cdc log table, and what exactly the "before" and "after" outputs look like.

I will. Thank you for your feedback!

nyh commented:
Can you please help me understand:
1. Before this patch, does "desc table" on a base table (table that has CDC enable) work? Does it produce acceptable results? If the answer is yes, why is that issue still open?
2. If this patch is about desc table on a **log table**, why do we care? Who will ever want to desc table on a log table? How does a user even know the log table's name? The user in that issue certainly didn't care about this, right?

dawmd commented:
> Can you please help me understand:
> 
>     1. Before this patch, does "desc table" on a base table (table that has CDC enable) work? Does it produce acceptable results? If the answer is yes, why is that issue still open?

Yes. As for why the issue is open, my assumption was the fact that the feature was still incomplete: although we could describe the base table, we couldn't describe the log table. But I get what you mean and I agree that the issue itself is confusing and didn't mention it explicitly indeed. Another reason that I see is a relatively long discussion regarding CDC log tables was held there (since it was the relevant context) and it makes sense to "connect" it somehow to a PR fixing the problem (the one with log tables).

>     2. If this patch is about desc table on a **log table**, why do we care? Who will ever want to desc table on a log table? 

It's for completeness of the feature. User shouldn't modify system tables (at least some of them) directly either, but we describe them when using `DESC FULL SCHEMA` anyway. Log tables are already included in the output of `DESC SCHEMA` too, albeit in the form of an ALTER statement. It's a fair assumption to think that you can also describe a log table directly.

> How does a user even know the log table's name?

We specify it in our [user documentation](https://opensource.docs.scylladb.com/stable/using-scylla/cdc/cdc-log-table.html).

> The user in that issue certainly didn't care about this, right?

I agree

nyh commented:
> Yes. As for why the issue is open, my assumption was the fact that the feature was still incomplete: although we could describe the base table, we couldn't describe the log table.

But the person who created this issue seemed to report a bug that describing the base table did not work. Was this bug ever fixed? Did we ever checked that bug even existed? Or if it was fixed, why wasn't the bug reported told that - basically his bug was solved? Actually, when was it solved? It really doesn't look to me like this patch is what is solving it.

Can you please try to figure out when the original user's bug report (that "desc table" for base table) got fixed and close the original issue manually - not in a "Fixes" in your new patch?
Then, if you think that the "completeness of the feature" requires us to be a able to explicitly "desc table" a hidden CDC table, please open a new issue about that, and explaining the reasoning why you think it's needed for completeness of the picture.

By the way, how does DESC SCHEMA/TABLE/etc work for other types of hidden tables? Another example I can think of are the materialized views created for secondary indexes. These views are never created directly - you "CREATE INDEX" and that creates the materialized view with some hidden name like tblname_index_a_idx or something. You certainly can't call both the CREATE INDEX and then CREATE MATERIALIZED VIEW. The question that is relevant for this issue is - can you "DESC TABLE" (or "DESC MATERIALIZED VIEW"?) such a "hidden" view?

dawmd commented:
> But the person who created this issue seemed to report a bug that describing the base table did not work. Was this bug ever fixed? Did we ever checked that bug even existed? Or if it was fixed, why wasn't the bug reported told that - basically his bug was solved? Actually, when was it solved? It really doesn't look to me like this patch is what is solving it.
>  
> Can you please try to figure out when the original user's bug report (that "desc table" for base table) got fixed and close the original issue manually - not in a "Fixes" in your new patch? Then, if you think that the "completeness of the feature" requires us to be a able to explicitly "desc table" a hidden CDC table, please open a new issue about that, and explaining the reasoning why you think it's needed for completeness of the picture.

Looking at the issue and the version of Scylla the user provided, it seems they didn't have access to server-side describe introduced in scylladb/scylladb#11106. Chances are it was never fixed in cqlsh, but the bug was never present in Scylla itself. That would explain why no PR ever referenced it. Perhaps @Jadw1 knows more context.

I'll wait for Michał's feedback and if we don't figure out what else might've been the case, I'll close the issue. In the meantime, I'll create a new one as you suggested.

> By the way, how does DESC SCHEMA/TABLE/etc work for other types of hidden tables? Another example I can think of are the materialized views created for secondary indexes. These views are never created directly - you "CREATE INDEX" and that creates the materialized view with some hidden name like tblname_index_a_idx or something. You certainly can't call both the CREATE INDEX and then CREATE MATERIALIZED VIEW. The question that is relevant for this issue is - can you "DESC TABLE" (or "DESC MATERIALIZED VIEW"?) such a "hidden" view?

I wasn't aware you could create an MV for a secondary index, but I'll try to test it and provide some information here later. That's certainly an interesting edge case.

nyh commented:
When you use CREATE INDEX behind the scenes a materialized view is created for you. You don't need to create this view yourself.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12671/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12671/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4379/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12401/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1311/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12671/)
#### Build Details:
- Duration: 7 hr 27 min
- Builder: i-0c13e0468d1d69ec0 (m5ad.8xlarge)

piodul commented:
@dawmd I have created a separate issue for tracking the problem of DESCRIBE being forbidden on CDC tables (https://github.com/scylladb/scylladb/issues/21235). Let's use it instead of #13702 - please update the description and commit messages.

dawmd commented:
Version 2:
* Added an example before and after the change.
* Changed the issue number to #21235

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12764/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12764/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12451/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1361/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4429/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12764/)
#### Build Details:
- Duration: 4 hr 16 min
- Builder: i-0059c8b1e595eddef (m5d.12xlarge)

piodul commented on cql3/statements/describe_statement.cc:
```diff
@@ -293,6 +289,17 @@ future<std::vector<description>> table(const data_dictionary::database& db, cons
         }
     }
 
+    if (cdc::is_log_for_some_table(db.real_database(), ks, name)) {
+        if (result.size() != 1) {
+            on_internal_error(dlogger, "CDC log table is not supposed to have secondary indexes or MVs derived from it");
+        }
+
+        // If the table the user wants to describe is a CDC log table, we want to print it as a CQL comment.
+        // This way, the user learns about the internals of the table, but it's implicitly implied that
+        // they should not execute the statement.
+        result[0].create_statement = seastar::format("/*\n{}\n*/", *result[0].create_statement);
+    }
```
I think there is a better way to write it. You can change the code after the `// table` comment:

```c++
    // table
    replica::schema_describe_helper describe_helper{db};
    auto base_description = schema->describe(describe_helper, with_internals ? describe_option::STMTS_AND_INTERNALS : describe_option::STMTS);
    if (cdc::is_log_for_some_table(db.real_database(), ks, name)) {
        // Transform the `base_description` accordingly and push it to `results`
    } else {
        result.push_back(std::move(base_description));
    }
```

This way, the code will not rely on the assumption that `result.size() == 1` here for cdc tables (you still rely on it, even if you verify it with an assert).

dawmd replied:
Good idea

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -970,6 +972,131 @@ def test_hide_cdc_table(scylla_only, cql, test_keyspace):
+        assert row.type == "table"
+        assert row.name == log_table_name
+        
+        # The actual content of `row.create_statement` is tested separately in `test_describe_cdc_log_table_layout`.
```
Did you mean `test_describe_cdc_log_table_create_statement`? I cannot find `test_describe_cdc_log_table_layout`.

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -970,6 +972,131 @@ def test_hide_cdc_table(scylla_only, cql, test_keyspace):
+        create_statement = row.create_statement
+
+        # We want to get rid of the CQL comment markers: "/*" and "*/", which make for the first and the last lines.
+        create_statement = create_statement.replace("/*\n", "")
+        create_statement = create_statement.replace("\n*/", "")
```
Nit: use `removeprefix` / `removesuffix`. This will make the intention of the code self-explanatory.

https://docs.python.org/3.9/library/stdtypes.html?highlight=removeprefix#str.removeprefix

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -970,6 +972,131 @@ def test_hide_cdc_table(scylla_only, cql, test_keyspace):
+            # All of that for the convenience of comparing the strings later on.
+
+            # We want to get rid of the CQL comment markers: "/*" and "*/", which make for the first and the last lines.
+            create_statement = create_statement.replace("/*\n", "")
+            create_statement = create_statement.replace("\n*/", "")
```
Same thing about removeprefix and removesuffix.

piodul commented on test/cql-pytest/test_describe.py:
```diff
@@ -51,6 +51,12 @@ def extract_names(desc_result_iter: Iterable[DescRowType]) -> Iterable[str]:
 
+###
+
+def get_lines(string: str, begin: int, end: int) -> str:
+    lines = string.split("\n")
+    return "\n".join(lines[1:-1])
```
Is this function used anywhere?

dawmd commented:
Version 3:
* Elaborated on why we display the create statement as a comment in the dev doc article.
* Reorganised the logic in `cql3/statements/describe_statement.cc`.
* Started using `remove{prefix, suffix}` in the tests instead of doing the same by hand; fixed indentation.
* Corrected a comment in one of the tests.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12824/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12824/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4477/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12499/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1410/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12824/)
#### Build Details:
- Duration: 3 hr 34 min
- Builder: spider1.cloudius-systems.com

piodul commented on docs/dev/describe_schema.md:
```diff
@@ -91,6 +91,43 @@ would require changing the configuration to be even able to insert the data. Tha
+The reason why `DESC TABLE <log table>` prints a `CREATE` and not an `ALTER` statement is that it enforces providing
+a list of its columns with their types, i.e. we provide complete information about the table.
+
+The rationale for printing the statement as a comment is that we want to implicitly suggest
+that the user should *not* execute the statement: the log table should only be created as a side effect of creating
+the base table, and describing a log table servers an informational purpose only.
```
Sorry, it looks like my previous comment about adding 1-2 sentences was missing one critical thing: I meant putting these sentences _in the comment that is produced by DESCRIBE_ - that's what I had in mind about being explicit rather than implicit, we want to explicitly tell the user why CREATE statement is wrapped in the comment markers and not try to implicitly suggest anything.

dawmd replied:
Ah, that makes more sense indeed. I'll adjust it in a moment

dawmd commented:
Version 4:
* The user is now explicitly informed in the comment that they shouldn't execute the returned statement.

dawmd commented:
Please, do NOT merge this PR until scylladb/scylladb#21352 has been merged to not introduce any conflicts. That PR has a much higher priority.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12855/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/15/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12535/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4509/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1439/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12855/)
#### Failed Tests (2/36607):
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12855/testReport/junit/boost.memtable_test.test_exceptions_in_flush_on_sstable_write.3.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
- [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12855/testReport/junit/boost.memtable_test.test_exceptions_in_flush_on_sstable_open.1.xunit/xml/Tests___Unit_Tests____empty_) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+[empty])
#### Build Details:
- Duration: 3 hr 32 min
- Builder: spider6.cloudius-systems.com

nyh commented:
> Please, do NOT merge this PR until #21352 has been merged to not introduce any conflicts. That PR has a much higher priority.

I merged that, so please rebase this PR (it should be easier for you if you wait until #21352 is promoted to master).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12945/)
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/71/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4587/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1514/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12612/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12945/)
#### Build Details:
- Duration: 7 hr 21 min
- Builder: i-0d28802c304bd97e3 (m5ad.8xlarge)

piodul commented:
Queued
